docs/rule/sort-invocations.md

```diff
@@ -0,0 +1,97 @@
+# sort-invocations
+
+🔧 The `--fix` option on the command line can automatically fix some of the problems reported by this rule.
+
+## Why use it?
+
+The rule helps you standardize templates:
+
+- Component invocations
+- Helper invocations
+- Modifier invocations
+
+By sorting things that are order-independent, you can more easily refactor code. In addition, sorting removes style differences, so you can review another person's code more effectively.
+
+> [!TIP]
+>
+> The `--fix` option for this rule doesn't preserve formatting. To preserve formatting, you can use [`ember-template-lint-plugin-prettier`](https://github.com/ember-template-lint/ember-template-lint-plugin-prettier) and [`prettier-plugin-ember-template-tag`](https://github.com/ember-tooling/prettier-plugin-ember-template-tag) to format `*.hbs` and `*.{gjs,gts}`, respectively.
+
+## Examples
+
+### Components
+
+When invoking a component, list things in the following order:
+
+1. Arguments
+2. Attributes
+3. Modifiers
+4. Splattributes
+
+The order clearly shows how the component is customized more and more. Things are alphabetized within each group.
+
+```hbs
+<Ui::Button
+  @label="Submit form"
+  @type="submit"
+  data-test-button
+  {{on "click" this.doSomething}}
+  ...attributes
+/>
+```
+
+### Helpers
+
+When invoking a helper, list the named arguments in alphabetical order.
+
+```hbs
+{{t
+  "my-component.description"
+  installedOn=this.installationDate
+  packageName="@ember/source"
+  packageVersion="6.0.0"
+}}
+```
+
+### Modifiers
+
+Similarly to helpers, list the named arguments in alphabetical order.
+
+## Limitations
+
+It's intended that there are no options for sorting. Alphabetical sort is the simplest for everyone to understand and to apply across different projects. It's also the easiest to maintain.
+
+To better meet your needs, consider creating a plugin for `ember-template-lint`.
+
+## Known issues
+
+1\. If you passed an empty string as an argument's value, it has been replaced with the placeholder text `<EMPTY STRING>`. Do a find-and-replace-all to restore the value.
+
+```diff
+- <MyComponent @description="<EMPTY STRING>" />
++ <MyComponent @description="" />
+```
+
+2\. Comments such as `{{! @glint-expect-error }}` may have shifted. Move them to the correct location.
+
+3\. If you passed a concatenated string that has exactly 1 character before or after a "mustache," you will need to add back the missing character.
+
+```diff
+<MyComponent
+  @isOpen={{this.isOpen}}
+-   @parentContainerId="{{@parentId}}"
++   @parentContainerId="#{{@parentId}}"
+/>
+```
+
+Better yet, use the `{{concat}}` helper instead (this helps codemods).
```

Comment:
I think we could manually do it, applying transform, because `concat` is builtin helper - https://api.emberjs.com/ember/3.2/classes/Ember.Templates.helpers/methods/get?anchor=concat

Reply:
When writing my codemod, I found that I'd need to recreate the `ConcatStatement`. Because we can pass more complex expressions to a concatenated string, I don't think we can easily recreate it without increasing the maintenance cost.

Here, it's better to ask end-developers to refactor code.

Reply:
We could have something like this - https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/rules/no-curly-component-invocation.js#L60

Reply:
`Here, it's better to ask end-developers to refactor code.` - I agree, but, we should produce kinda "safe" code in fixes, in worst case, I would prefer to not sort empty attribute and print meaningful message (saying where it to place and that we can't automatically adjust it)

Reply:
Code that will run into this issue should occur rarely (it happened twice in a large production project, and both were caught thanks to tests). So I think this rule shouldn't handle this rare case.

A more maintainable approach would be to write a rule in `ember-template-lint` that encourages people to use the `{{concat}}` helper instead.

Reply:
wondering if we skip this element/component node at all, if issue is rare, it's better to just emit error, fix others and disable autofix only for this case

Reply:
I came up with a fix in https://github.com/ijlee2/ember-codemod-sort-invocations/pull/13/files. I'll downstream it soon.